[ray.data.llm] Fix build_llm_processor for ServeDeploymentProcessor

jeffreyjeffreywang opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
`chat_template_kwargs` was introduced in #56490 and is currently passed through all processors, even though some (e.g., `ServeDeploymentProcessor` and `HttpRequestProcessor`) do not support this field. As a result, constructing a `ServeDeploymentProcessor` via `build_llm_processor` raises errors.

This issue wasn’t caught in tests because they have been using `ProcessorBuilder.build`, which bypasses the layer where `chat_template_kwargs` is passed.

In this PR, we:
- Introduce `builder_kwargs` and delegate their validation to individual processors. The `chat_template_kwargs` will be passed as part of `builder_kwargs` to the `vLLMEngineProcessor`.
- Update tests to rely on public APIs rather than developer APIs, ensuring they reflect end-user usage.

## Related issue number

<!-- For example: "Closes #1234" -->
Resolves #57060 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run pre-commit jobs to lint the changes in this PR. ([pre-commit setup](https://docs.ray.io/en/latest/ray-contribute/getting-involved.html#lint-and-formatting))
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Make build_llm_processor skip chat_template_kwargs for non-offline processors and update processor tests to use public ray.data.llm APIs.
> 
> - **Core (ray.data.llm)**
>   - Adjust `build_llm_processor` to only pass `chat_template_kwargs` when `config` is an `OfflineProcessorConfig`.
> - **Tests**
>   - Switch to public APIs: import processor configs from `ray.data.llm` and use `build_llm_processor` instead of `ProcessorBuilder.build` where applicable.
>   - Update Serve Deployment and vLLM processor tests accordingly; keep behavior/assertions intact.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit fb36cc3df97e37a8abc76dca03cdd07fea24cec2. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

jeffreywang88 committed:
build_llm_processor should not pass chat_template_kwargs to processors that don't support it

Signed-off-by: jeffreyjeffreywang <jeffjeffreywang@gmail.com>

jeffreywang88 committed:
Apply gemini feedback

Signed-off-by: jeffreyjeffreywang <jeffjeffreywang@gmail.com>

jeffreyjeffreywang commented:
cc: @nrghosh

nrghosh commented:
cc @ahao-anyscale

nrghosh reviewed (commented):
cc @ahao-anyscale

nrghosh reviewed (approved):
ltgm pending tests

kouroshHakha reviewed (commented):
There is a fundamental problem found in this issue / PR that needs to be addressed properly. The gist of the issue is that we did not enforce any standard signature for builder methods. it implicitly started with `config, preprocess, postprocess` and for some builders we added chat_template_kwargs` that went missing by other builders that were not on the critical path. The proper fix is either: 1) enforce a strong typing / signature for the builder methods or 2) make the builder kwargs a passthrough towards the registered builder functions. 

So here is my proposal for fixing this fundamentally and is more aligned with approach 2.

1. change the signature of the public `build_llm_processor` to (config, preprocess, postprocess, builder_kwargs)

2. Keep the old things as is, and pass through the builder_kwargs to the registered builder function. If the builder function does not have the matching signature it will barf during the build.



The current fix in this PR is not good, because it creates a leakage that OfflineProcessorConfigs will always have chat_template_kwargs which is again an implicit assumption. 


I think the second approach is more generic and flexible.

nrghosh committed:
address comments

- use generic `builder_kwargs` passthrough in `build_llm_processor`
- replace `chat_template_kwargs` param with `builder_kwargs` dict that
passes through to the registered builder functions
- remove implicit assumptions and have builders raise `TypeError` for
unsupported kwargs
- test passthrough with conflicts

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh reviewed (commented):
taking a first pass to address comments

hao-aaron reviewed (commented):
Agree with kourosh, i think it makes more sense for arg checking to be done inside each builder function.

jeffreyjeffreywang commented:
microcheck failure doesn't seem to be relevant:
```
[2025-10-03T21:12:09Z] --------
  | [2025-10-03T21:12:09Z]  > [3/3] RUN <<EOF (#!/bin/bash -i...):
  | [2025-10-03T21:12:09Z] 62.85 + /ray/ci/env/../../ci/suppress_output conda install -c anaconda mpi4py -y
  | [2025-10-03T21:12:09Z] 63.08
  | [2025-10-03T21:12:09Z] 63.08 real	0m0.229s
  | [2025-10-03T21:12:09Z] 63.08 user	0m0.198s
  | [2025-10-03T21:12:09Z] 63.08 sys	0m0.030s
  | [2025-10-03T21:12:09Z] 63.08 Error while loading conda entry point: conda-libmamba-solver (cannot import name 'Spinner' from 'conda.common.io' (/opt/miniforge/lib/python3.9/site-packages/conda/common/io.py))
  | [2025-10-03T21:12:09Z] 63.08
  | [2025-10-03T21:12:09Z] 63.08 CondaValueError: You have chosen a non-default solver backend (libmamba) but it was not recognized. Choose one of: classic
  | [2025-10-03T21:12:09Z] 63.08
  | [2025-10-03T21:12:09Z] 63.08 FAILED 1
  | [2025-10-03T21:12:09Z]
```

jeffreyjeffreywang commented on python/ray/data/llm.py:
```diff
@@ -430,17 +434,58 @@ def build_llm_processor(
+            import ray
+            from ray.data.llm import vLLMEngineProcessorConfig, build_llm_processor
+
+            config = vLLMEngineProcessorConfig(
```
@nrghosh were you able to validate this example locally?

nrghosh replied:
yes, made it a smaller (but demonstrable) model so it's easier to validate that when you pass `enable_thinking = True` -> you find `<think> </think>` tags in the output, and not with `False` (that the kwargs are indeed passed properly).

jeffreyjeffreywang replied:
Thank you! I was able to run it with a A10G.

nrghosh committed:
run + update docstring example

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

kouroshHakha commented on python/ray/data/llm.py:
```diff
@@ -430,16 +434,65 @@ def build_llm_processor(
         postprocess=postprocess,
     )
 
+    # Pass through any additional builder kwargs
+    if builder_kwargs is not None:
+        # Check for conflicts with explicitly passed arguments
+        reserved_keys = {"preprocess", "postprocess"}
+        conflicting_keys = reserved_keys & builder_kwargs.keys()
+        if conflicting_keys:
+            raise ValueError(
```
could you create a utility on `ProcessBuilder` to do this validation there? 

```
ProcessBuilder.validate_builder_kwargs(builder_kwargs)

return ProcessBuilder.build(config, preprocess, postprocess, builder_kwargs)
```

jeffreyjeffreywang replied:
Sounds good, done in the latest revision.

kouroshHakha commented on python/ray/llm/tests/batch/cpu/processor/test_processor_base.py:
```diff
@@ -188,6 +188,110 @@ def overrider(name: str, stage: StatefulStage):
     )
 
 
+def test_builder_kwargs_passthrough():
```
Could you spearate out these test scenarios into different test methods? Maybe create a Test class to capture them all and share resource between them using fixtures, etc.

jeffreywang88 committed:
CR feedback

Signed-off-by: jeffreyjeffreywang <jeffjeffreywang@gmail.com>

jeffreywang88 committed:
Fix pydoclint

Signed-off-by: jeffreyjeffreywang <jeffjeffreywang@gmail.com>
